src/entities.rs

```diff
@@ -529,6 +532,45 @@ impl Entities {
     pub fn len(&self) -> u32 {
         self.len
     }
+
+    pub fn freelist(&self) -> impl ExactSizeIterator<Item = Entity> + '_ {
+        let free = self.free_cursor.load(Ordering::Relaxed);
+        let ids = match usize::try_from(free) {
+            Err(_) => &[],
+            Ok(free) => &self.pending[0..free],
+        };
+        ids.iter().map(|&id| Entity {
+            id,
+            generation: self.meta[id as usize].generation,
+        })
+    }
+
+    pub fn set_freelist(&mut self, freelist: &[Entity]) {
+        #[cfg(debug_assertions)]
+        {
+            for entity in freelist {
+                let Some(meta) = self.meta.get(entity.id as usize) else {
+                    continue;
+                };
+                assert_eq!(
+                    meta.location.index,
+                    u32::MAX,
+                    "freelist addresses live entities"
+                );
+            }
+        }
+        if let Some(max) = freelist.iter().map(|e: &Entity| e.id()).max() {
+            if max as usize > self.meta.len() {
```

Comment:
Should be `>=` to avoid out of bounds array access at `self.meta[entity.id as usize].generation = entity.generation;` below.

Minimal test case you might or might not want to add:

```rust

    #[test]
    fn test_set_freelist_causes_resize() {
        let mut world = World::new();
        let a = world.spawn(());
        world.despawn(a).unwrap();
        let freelist = world.freelist().collect::<Vec<_>>();

        let mut world2 = World::new();
        world2.set_freelist(&freelist);
    }
```

Reply:
Great catch, thanks!